phtree/common/filter.h

```diff
@@ -150,6 +150,132 @@ class FilterAABB {
+     * means that the paraboloid opens in the positive direction of the axis, a negative
+     * scale factor means the reverse.
+     */
+    FilterParaboloid(
+        const size_t axis,
+        const KeyExternal& vertex,
+        const ScalarExternal& scale_factor,
+        CONVERTER converter = CONVERTER(),
+        DISTANCE distance_function = DISTANCE())
+    : axis_{axis}
+    , vertex_external_{vertex}
+    , vertex_internal_{converter.pre(vertex)}
+    , scale_factor_{scale_factor}
+    , converter_{converter}
+    , distance_function_{distance_function} {};
+
+    template <typename T>
+    [[nodiscard]] bool IsEntryValid(const KeyInternal& key, const T& value) const {
+        if (
+            (key[axis_] > vertex_internal_[axis_] && scale_factor_ < 0) ||
+            (key[axis_] < vertex_internal_[axis_] && scale_factor_ > 0)) {
+            // point is on the side of the vertex where the paraboloid does not exist
+            return false;
+        }
+
+        KeyExternal point = converter_.post(key);
+
+        // radius of paraboloid at axis offset of point
+        ScalarExternal diff_axis = point[axis_] - vertex_external_[axis_];
+        ScalarExternal radius = sqrt(diff_axis / scale_factor_);
+
+        // point projected onto axis of paraboloid
+        KeyExternal projected = vertex_external_;
+        projected[axis_] = point[axis_];
+
+        return distance_function_(projected, point) <= radius;
+    }
+
+    /*
+     * Calculate whether AABB encompassing all possible points in the node intersects with the
+     * paraboloid.
+     */
+    [[nodiscard]] bool IsNodeValid(const KeyInternal& prefix, int bits_to_ignore) const {
+        // we always want to traverse the root node (bits_to_ignore == 64)
+
+        if (bits_to_ignore >= (MAX_BIT_WIDTH<ScalarInternal> - 1)) {
+            return true;
+        }
+
+        ScalarInternal node_min_bits = MAX_MASK<ScalarInternal> << bits_to_ignore;
+        ScalarInternal node_max_bits = ~node_min_bits;
+
+        // coordinate on axis that is inside the paraboloid and furthest away from vertex
+        ScalarInternal coord_max_vertex_dist;
+        if (scale_factor_ < 0) {
+            // lower bound is furthest from the vertex
+            coord_max_vertex_dist = prefix[axis_] & node_min_bits;
+            if (coord_max_vertex_dist > vertex_internal_[axis_]) {
+                // point is on the side of the vertex where the paraboloid does not exist
+                return false;
+            }
+        } else {
+            // upper bound is furthest from the vertex
+            coord_max_vertex_dist = prefix[axis_] | node_max_bits;
+            if (coord_max_vertex_dist < vertex_internal_[axis_]) {
+                // point is on the side of the vertex where the paraboloid does not exist
+                return false;
+            }
+        }
+
+        KeyInternal closest_to_axis;
+        for (size_t i = 0; i < prefix.size(); ++i) {
```

Comment:
Nit: this could use `3` or maybe the DIM provided by the converter, instead of `.size()`

Reply:
Done. Note that `FilterAABB` and `FilterSphere` both use `.size()`. Doing it in this PR does not make sense as it's not clear if it will be merged.

Reply:
Thanks for pointing that out. I think `FilterAABB` and `FilterSphere` had inconsistent use of DIM vs size(), I fixed that.